[REFACTOR] Add exception handler

sunsun512 opened the pull request:
## 변동사항

> 기존 Exception은 상태 코드 400이상과 함께 Response를 반환하여 사용하는 다른 시스템과 내부 로그에서 문제를 파악하기 한계가 있다.
-  Exception을 custom해서 상황 별로 특정한 코드와 메시지를 반환하고, 다른 서비스에서 호출 해도 HTTP status == ok 와 에러 원인을 알 수 있도록   Exception handler를 추가하였습니다.

sunsun512 committed:
Merge pull request #2 from sunsun512/dividing-controller

Dividing controller

sunsun512 committed:
feat: BusinessException 추가
- 서비스 내 에러 BusinessException 추가

sunsun512 committed:
feat: BdExceptionHandler 추가
- 서비스 response 형식 추가
- BusinessException 코드와 메시지 추가
- BusinessException 에러 시 response ok로 반환

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/error/BusinessException.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+
+}
```
fyi. record

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/UrlShortenService.java:
```diff
@@ -17,19 +19,21 @@ public class UrlShortenService {
     }
 
     public String shortenUrlCreate(String originUrl){
+        //@todo url validation 필요
```
혹시 todo 주석에 @ 를 넣는건 어디에서 보신건가요?
저는 그냥 todo 로만 달고 있는데, @ 넣는건 처음봐서요.

fyi. https://www.jetbrains.com/help/idea/using-todo.html

sunsun512 replied:
@Hyune-c 저도 다른 분이 쓰는 거 보고 알게 됬는데, 그냥 문자로도 쓰면 되는군요! 몰랐습니다. ㅎ

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/UrlShortenService.java:
```diff
@@ -17,19 +19,21 @@ public class UrlShortenService {
+        //@todo url validation 필요
         if (!StringUtils.hasText(originUrl)) {
-            throw new IllegalArgumentException("Request URL No exists ");
+            throw BusinessExceptionGenerator.createBusinessException("DB001", "Request URL No exists ");
```
error_code 에 대해서는 갑론을박이 있습니다.
- 가시성이 떨어진다.
- 규격화가 되서 좋다. 등등

장단점이 있는 것이라 어떤 형태를 해도 좋지만 지금과 같이 낱개 String 은 위험 합니다.

fyi. 매직 넘버

Hyune-c replied:
ErrorCode 를 정의해놓으신게 있네요? 그것을 사용하시면 좋을 것 같습니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/error/BusinessExceptionGenerator.java:
```diff
@@ -0,0 +1,35 @@
+
+@Slf4j
+@Component
+public class BusinessExceptionGenerator {
```
이 클래스는 팩토리 성격인 것 같은데, 굳이 component 여야 되나요?

sunsun512 replied:
이는 Spring Framework에서 해당 클래스를 컴포넌트로 등록하여 Spring IoC 컨테이너가 관리할 수 있도록 하는 것입니다.

일반적으로 factory 성향이 강한 클래스는 @Component를 사용하여 Spring의 관리 대상으로 만드는 경우가 많습니다. 이렇게 함으로써 다음과 같은 이점이 있습니다:

의존성 주입(Dependency Injection) 관리: @Component로 선언된 클래스는 Spring IoC 컨테이너에 의해 관리되며, 필요한 곳에서 해당 클래스의 인스턴스를 주입받아 사용할 수 있습니다. 예를 들어, BusinessExceptionGenerator 클래스가 다른 컴포넌트에서 사용되어야 할 때, Spring이 자동으로 필요한 인스턴스를 제공해줍니다.

싱글톤(Singleton) 스코프 관리: 기본적으로 Spring의 컴포넌트는 싱글톤으로 관리됩니다. 따라서 애플리케이션 전역에서 하나의 인스턴스만 생성되어 공유되며, 메모리 사용을 줄이고 성능을 최적화할 수 있습니다.

편리한 테스트: @Component로 선언된 클래스는 Spring의 테스트 환경에서 쉽게 모의(mock) 객체로 대체할 수 있어 유닛 테스트를 수행하기 용이합니다.

따라서, BusinessExceptionGenerator 클래스가 factory 메서드를 제공하면서도 Spring의 IoC/DI 기능을 활용하기 위해 @Component 어노테이션을 사용한 것입니다.

> 참고하기 위한 chatGPT 답변입니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/error/RestControllerAdvice.java:
```diff
@@ -0,0 +1,64 @@
+                        .build());
+    }
+
+}
```
왜 이 클래스만 lint 가 망가졌을까요?

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/error/RestControllerAdvice.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    } catch (Exception e) {
+        log.error("Error while extracting exception details", e);
+    }
```
`nit` 꽤 모호한 try-catch 네요..

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/util/UrlShortenUtil.java:
```diff
@@ -0,0 +1,25 @@
+        return ResponseEntity.ok(
+                ResultJson.builder().data(convert).build());
+    }
+}
```
fyi. https://mattilehtinen.com/articles/dunghill-anti-pattern-why-utility-classes-and-modules-smell/

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/util/model/ResultJson.java:
```diff
@@ -0,0 +1,39 @@
+    @Builder.Default
+    Long timeStamp = System.currentTimeMillis();
+
+}
```
`nit` 공용으로 쓰기 위한 `표준 응답 객체` 를 만드신거 같군요. 
개인적으로는 안좋아하는 스타일이긴 하지만 많이 쓰이는 방법이긴 합니다.

sunsun512 replied:
@Hyune-c 혹시 선호하시는 스타일은 어떤 스타일인지 여쭤봐도 될까요?

sunsun512 committed:
feat: comment 반영
- RestControllerAdvice - extractExceptionDetails 사용 안함으로 삭제
- UrlShortenService 내에서 에러 코드만 사용
- UrlShortenUtil에서 ResponseFormatter로 동작이 보이는 이름으로 변경
